### PR TITLE
fix: remove double instrumentation of mongob commands

### PR DESCRIPTION
**Pull request by david-luna**
While doing investigations for #3161 some added logs revealed we were creating two spans for each command it was started. The root cause of this was because the instrumentation was registering the event listeners twice and as a result
- 2 spans where created
- the 2nd span overwrites the 1st on the map of active spans
- the 2nd span is ended but not the 1st

Since we are patching the `MongoClient#constructor` and also the static `connect` method whenever a user was calling `connect` to get a client this happens:
- the `patched connect` calls its original function
- the `original connect` method creates a new instance using `new MongoClient(...)`
- the `patched constructor` of MongoClient adds listeners
- the `patched connect` adds listeners to the client returned by its original function (duplication)

In conclusion there is no need to patch `connect` if already adding listeners in the constructor.

NOTE: this behavior started to happen from v14.11.0 when the creation of the instance changed from `new MongoClient(...)` to `new this(...)` making the patched constructor to run

### Checklist

- [x] Implement code
- [ ] Add tests
- [ ] Update TypeScript typings
- [ ] Update documentation
- [x] Add CHANGELOG.asciidoc entry
- [x] Commit message follows [commit guidelines](https://github.com/elastic/apm-agent-nodejs/blob/main/CONTRIBUTING.md#commit-message-guidelines)

### CODE REVIEW

**Review by david-luna — COMMENTED** [review]
/test tav mongodb

**Review by david-luna — COMMENTED** [review]
/test tav mongodb

**Review by david-luna — COMMENTED** [review]
/test tav mongodb

**Review by david-luna — COMMENTED** [review]
/test tav mongodb

File: CHANGELOG.asciidoc [hunk 1]
```diff
@@ -38,6 +38,11 @@ Notes:
 [float]
 ===== Bug fixes
 
+* Fix instrumentation of `mongodb` to avoid multiple command handler
+  registrations when client is created via `MongoClient.connect` static
+  method. ({pull}3586[#3586])
+
+
```
Comment (trentm) [inline CHANGELOG.asciidoc:45]: This should be added to a new `        ==== Unreleased` block.

File: test/_spy_method.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+/**
+ * Returns an object with data about the usage of the spied method
+ *
+ * @param {any} test the test where the spy is created
+ * @param {Object} target the object which holds the method to spy
+ * @param {String} method the method to spy on
+ * @returns {SpyObject} object with usage data
+ */
+function spyOn(test, target, method) {
```
Comment (trentm) [inline test/_spy_method.js:38]: nit: I would rename this `test` to `t`, which is the typical var name used in tape and tap tests.

File: test/_spy_method.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    throw Error('spyOn: target or method not defined');
+  }
+  if (!target[method]) {
+    throw Error(`spyOn: method ${method} is not defined on target`);
+  }
+  if (typeof target[method] !== 'function') {
+    throw Error(`spyOn: method ${method} is not a function`);
+  }
+  if (typeof target[method][kOrigFunction] === 'function') {
+    throw Error(`spyOn: method ${method} is already used`);
```
Comment (trentm) [inline test/_spy_method.js:49]: nit: perhaps this?

```suggestion
    throw Error(`spyOn: method ${method} is already being spied on`);
```

File: test/instrumentation/modules/mongodb.test.js [hunk 4]
```diff
@@ -206,6 +216,11 @@ test('ensure run context', async function (t) {
         `span ${s.type}.${s.subtype} "${s.name}" is a child of the transaction`,
       );
     });
+    // Internal calls to `createSpan` are not duplicated
+    t.ok(
+      createSpanSpy.hasBeenCalled(3),
+      `spans created: expected 3, actual ${createSpanSpy.calls.length}`,
```
Comment (trentm) [inline test/instrumentation/modules/mongodb.test.js:222]: nit: I was a bit confused as to why this is 3 and not 4.  It is because `agent.startSpan('manual').end();` doesn't happen to call `agent._instrumentation.createSpan()`, because of an implementation detail.

This is totally optional. A potential alternative would be to use the existing `transaction.span_count.started` value reported on serialized transactions. Like this: https://gist.github.com/trentm/032e3b560dc8ce4c3d1092253a956890

Then `spyOn` wouldn't be needed in this case. I think `spyOn` is still a useful test utility that could be added here, or saved for a separate commit.

**Review by trentm — CHANGES_REQUESTED** [review]
Nice find!  The changelog message position should move up. The rest of my comments are nits.

File: CHANGELOG.asciidoc (hunk 1 above)
Comment (trentm) [reply CHANGELOG.asciidoc:45]: If https://github.com/elastic/apm-agent-nodejs/pull/3590 goes in first and this merges in those changes, then this might be solved.

File: test/instrumentation/modules/mongodb.test.js (hunk 4 above)
Comment (david-luna) [reply test/instrumentation/modules/mongodb.test.js:222]: Oh, I did not know about `transaction.span_count.started`. If its just a counter and there is no internal logic that tweaks its value then I think its better to use that. I'll apply your suggestion, thanks

File: test/_spy_method.js (hunk 2 above)
Comment (david-luna) [reply test/_spy_method.js]: finally I decided to remove `spyOn` method since it's not necessary in this tests

File: test/instrumentation/modules/mongodb.test.js (hunk 4 above)
Comment (trentm) [reply test/instrumentation/modules/mongodb.test.js]: > Oh, I did not know about `transaction.span_count.started`.

Yah, totally fair. It exists for collecting data on dropped spans when hitting `transactionMaxSpans` limits, so it is a little indirect and lucky to be usable for this.

**Review by trentm — APPROVED** [review]
